src/jit/optimizer.cpp

```diff
@@ -7021,24 +7028,48 @@ void Compiler::optHoistLoopBlocks(unsigned loopNum, ArrayStack<BasicBlock*>* blo
             // If this 'tree' is hoistable then we return and the caller will
             // decide to hoist it as part of larger hoistable expression.
             //
-            if (!treeIsHoistable)
+            if (!treeIsHoistable && treeHasHoistableChildren)
             {
-                // We are not hoistable so we will now hoist any hoistable children.
+                // The current tree is not hoistable but it has hoistable children that we need
+                // to hoist now.
+                //
+                // In order to preserve the original execution order, we also need to hoist any
+                // other hoistable trees that we encountered so far.
+                // At this point the stack contains (in top to bottom order):
+                //   - the current node's children
+                //   - the current node
+                //   - previously traversed nodes (descendants of some current node's ancestor)
```

Comment:
Won't it contain the current node parent and its ancestors as well?
They won't have `value.m_hoistable` but I think they are on the stack now, because we pushed them `PreOrderVisit` and haven't poped yet.

Reply:
Yes, that's practically included in "- previously traversed nodes (descendants of some current node's ancestor)" but it's probably not 100% clear/accurate (e.g. the root of the tree is there too but it's not the descendant of anything). I'll try to improve the comment, it should probably simply say that at this point the stack contains already visited nodes and some of them may be hoistable.

And of course, `m_hoistable` would be `false` for ancestors since it defaults to `false` when the value is pushed on the stack.

I suppose this value stack thing is a bit weird but it's a cheap way to associate information with tree nodes. Otherwise we'd traverse the linear order and put the "hoistable" information in the tree node itself or in some sort of `GenTree*` -> `Value` hashmap.